crates/bevy_utils/src/label.rs

```diff
@@ -66,53 +106,114 @@ macro_rules! define_label {
             /// Converts this type into an opaque, strongly-typed label.
+            fn as_label(&self) -> $id_name;
+        }
+
+        $(#[$into_attr])*
+        pub trait $into_label: 'static {
+            /// Returns a number used to distinguish different labels of the same type.
+            fn data(&self) -> u64;
+            /// Writes debug info for a label of the current type.
+            /// * `data`: the result of calling [`data()`](#method.data) on an instance of this type.
+            fn fmt(data: u64, f: &mut ::std::fmt::Formatter) -> ::std::fmt::Result;
+        }
+
+        impl<T: $into_label> $label_name for T {
+            #[inline]
             fn as_label(&self) -> $id_name {
-                let id = self.type_id();
-                let label = self.as_str();
-                $id_name(id, label)
-            }
-            /// Returns the [`TypeId`] used to differentiate labels.
-            fn type_id(&self) -> ::core::any::TypeId {
-                ::core::any::TypeId::of::<Self>()
+                // This is just machinery that lets us store the TypeId and formatter fn in the same static reference.
+                struct VTables<L: ?::std::marker::Sized>(L);
+                impl<L: $into_label + ?::std::marker::Sized> VTables<L> {
+                    const VTABLE: $crate::label::VTable = $crate::label::VTable {
+                        ty: || ::std::any::TypeId::of::<L>(),
+                        fmt: <L as $into_label>::fmt,
+                    };
+                }
+
+                let data = self.data();
+                $id_name { data, vtable: &VTables::<Self>::VTABLE }
             }
-            /// Returns the representation of this label as a string literal.
-            ///
-            /// In cases where you absolutely need a label to be determined at runtime,
-            /// you can use [`Box::leak`] to get a `'static` reference.
-            fn as_str(&self) -> &'static str;
         }
 
         impl $label_name for $id_name {
+            #[inline]
             fn as_label(&self) -> Self {
                 *self
             }
-            fn type_id(&self) -> ::core::any::TypeId {
-                self.0
+        }
+
+        impl $id_name {
+            /// Returns the [`TypeId`] of the label from which this ID was constructed.
+            ///
+            /// [`TypeId`]: ::std::any::TypeId
+            #[inline]
+            pub fn type_id(self) -> ::std::any::TypeId {
+                (self.vtable.ty)()
+            }
+            /// Returns true if this label was constructed from an instance of type `L`.
+            pub fn is<L: $label_name>(self) -> bool {
+                self.type_id() == ::std::any::TypeId::of::<L>()
             }
-            fn as_str(&self) -> &'static str {
-                self.1
+            /// Attempts to downcast this label to type `L`.
+            ///
+            /// As an anti-footgun measure, the returned reference-like type is `!Send + !Sync`
```

Comment:
Very cool, but I'm not convinced that this is worth the added complexity. What failure cases are you trying to guard against in particular?

Reply:
I thought it might be surprising if the return value was sometimes `Send + Sync` and sometimes not (opaque types leak those implementation details), so I thought it would be more consistent if the reference was always `!Send + !Sync`.

I suppose it doesn't really prevent any footguns, since mutex guards already don't implement those traits.

Reply:
Yeah, IMO we just cut this complexity for now; we can refer back to this strategy if we find a compelling argument later.